ui/src/criteria/attribute.tsx

```diff
@@ -81,23 +133,145 @@ function AttributeEdit(props: AttributeEditProps) {
 
-  const selectionIndex = (hint: tanagra.EnumHintValue) =>
-    props.data.selected.findIndex(
-      (row) => row.id === hint.attributeValue?.int64Val
-    );
+  const hasValidIntegerBounds =
+    isValid(integerBoundsHint?.min) && isValid(integerBoundsHint?.max);
 
-  if (enumHintValues?.length === 0) {
+  if (!hasValidIntegerBounds && enumHintValues?.length === 0) {
     return (
       <Typography>
         No information for attribute {props.data.attribute}.
       </Typography>
     );
   }
+
+  if (hasValidIntegerBounds) {
+    const minBound = integerBoundsHint?.min || 0; // This is to ensure the compiler won't complain the object might be undefined
+    const maxBound = integerBoundsHint?.max || 0; // although we already know that min and max is valid
+
+    //two sets of values are needed due to the input box and slider is isolated.
+    const [minInputValue, setMinInputValue] = useState<number>(minBound);
+    const [maxInputValue, setMaxInputValue] = useState<number>(maxBound);
+    const [minValue, setMinValue] = useState<number>(minBound);
+    const [maxValue, setMaxValue] = useState<number>(maxBound);
+
+    const updateValue = (newMin: number, newMax: number) => {
+      setMinValue(newMin);
+      setMaxValue(newMax);
+      props.dispatchFn(
+        produce(props.data, (data) => {
+          data.min = newMin;
+          data.max = newMax;
+        })
+      );
+    };
+
+    const handleChange = (event: Event, newValue: number | number[]) => {
+      newValue = newValue as number[];
+      const [newMin, newMax] = newValue;
+      setMinInputValue(newMin);
+      setMaxInputValue(newMax);
+      updateValue(newMin, newMax);
+    };
+
+    const handleLeftInputChange = (
+      event: React.ChangeEvent<HTMLInputElement>
+    ) => {
+      let newMin =
+        event.target.value === "" ? minBound : Number(event.target.value);
+      setMinInputValue(newMin);
+
+      newMin = Math.max(minBound, newMin);
+      newMin = Math.min(maxValue, newMin);
+      updateValue(newMin, maxValue);
+    };
+    const handleLeftInputBlur = () => {
+      let newMin = minInputValue;
+      newMin = Math.max(newMin, minBound);
+      newMin = Math.min(newMin, maxValue);
+      setMinInputValue(newMin);
+    };
+
+    const handleRightInputChange = (
+      event: React.ChangeEvent<HTMLInputElement>
+    ) => {
+      let newMax =
+        event.target.value === "" ? maxBound : Number(event.target.value);
+      setMaxInputValue(newMax);
+
+      newMax = Math.max(minValue, newMax);
+      newMax = Math.min(newMax, maxBound);
+      updateValue(minValue, newMax);
+    };
+    const handleRightInputBlur = () => {
+      let newMax = maxInputValue;
+      newMax = Math.max(minValue, newMax);
+      newMax = Math.min(newMax, maxBound);
+      setMaxInputValue(newMax);
+    };
+
+    return (
+      <Box sx={{ width: "30%", minWidth: 100 }}>
+        <Grid container spacing={2} alignItems="center">
+          <Grid item>
+            <Input
```

Comment:
Do you mean I should set a maxlength="4" ?

Reply:
I'm not familiar with the parameters. Does that make the input large enough for 4 digits?

Reply:
Actually, that sounds like it would restrict it to only 4 digits which isn't really what we want. It just needs to be wider.

Reply:
I think right now(the current commit after this review) will support to see 4 digits. for example, when you go to the left input (minBound), backspace until empty. you can type  9 9 9 9. it is visible to see "9999", but when your mouse move and does not focus on the input. It will automatically changed to the minBound number due to 9999 is not supported.